docs/pallas/gpu/reference.md

```diff
@@ -162,6 +307,18 @@ There are three operations that can complete a barrier:
   Failing to satisfy this will corrupt the data structure and can cause surprising failures
   (including CUDA runtime errors). See below for an example of a valid program with two threads.
 
+> Another critical restriction is that the number of barrier completions must equal the
+  number of barrier waits throughout its lifetime. It is not allowed to end a scoped
+  allocation of a barrier when it has an unawaited completion. It can be subsequently
+  reused by the compiler and leaving it in this state can cause downstream problems.
+
+> Finally, it is crucial to ensure that each thread that ever waits on a `Barrier`
+  takes part in all `wait` operations on it. It is not allowed to e.g. await every
+  other completion of a barrier from one thread, and all other completions from another
+  one. Doing so will lead to deadlocks.
+
+  Note that the `Barrier` can receive arrivals from any source, without restrictions.
```

Comment:
Do you mean to say that several warpgroups can synchronize using the same barrier?

I'm not entirely sure what caveat we're trying to explain here/why this is important?

Reply:
I'm not sure how to explain it better than the docs do, but it is not allowed to do some waits on a barrier from one WG only, then from another WG, and then from the first WG. It is ok if they _always both wait_, but it is not ok when they _sometimes wait_.